### PR TITLE
Fix | Fixed GenerateSspiClientContext to retry negotiation with default port

### PR DESCRIPTION
**Pull request by arellegue**
This PR fixes Issue #2394 

When an application is targeting .Net 6.0, the application connects to the database successfully from Linux because the application is using System.Net.Security.NegotiateStreamPal.Unix.InitializeSecurityContext function which is retrying the negotiation with default port number 1433. However, when the application is targeting .Net8.0, the application fails to connect to the database from Linux because the newer implementation, MDS.SSPI.NegotiateSSPIContextProvider, did not retry with default port number. This issue is fixed in this PR.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SSPI/NegotiateSSPIContextProvider.cs [hunk 1]
```diff
@@ -14,13 +14,25 @@ internal sealed class NegotiateSSPIContextProvider : SSPIContextProvider
 
         internal override void GenerateSspiClientContext(ReadOnlyMemory<byte> received, ref byte[] sendBuff, ref uint sendLength, byte[][] _sniSpnBuffer)
         {
-            _negotiateAuth ??= new(new NegotiateAuthenticationClientOptions { Package = "Negotiate", TargetName = Encoding.Unicode.GetString(_sniSpnBuffer[0]) });
-            sendBuff = _negotiateAuth.GetOutgoingBlob(received.Span, out NegotiateAuthenticationStatusCode statusCode)!;
-            SqlClientEventSource.Log.TryTraceEvent("TdsParserStateObjectManaged.GenerateSspiClientContext | Info | Session Id {0}, StatusCode={1}", _physicalStateObj.SessionId, statusCode);
+            NegotiateAuthenticationStatusCode statusCode = NegotiateAuthenticationStatusCode.UnknownCredentials;
+
+            foreach (byte[] spn in _sniSpnBuffer)
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SSPI/NegotiateSSPIContextProvider.cs:19]: I'd rather not creating a new byte array variable and directly using `_sniSpnBuffer` inside the classic for-loop to avoid the extra stack operations.
```
for (int i = 0; i < _sniSpnBuffer.Length; i++)
{
    _sniSpnBuffer[i]; 
}
```

**Review by DavoudEshtehari — COMMENTED** [review]
@David-Engel Do you see any straight forward solution to add unit test in this case?

**Comment by David-Engel** [comment]
> Do you see any straight forward solution to add unit test for this case?

Straightforward, no.  It would be an edge case that would require specific environment setup and conditional checking on the test. I'd rather just get the fix in than make it wait for a test.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SSPI/NegotiateSSPIContextProvider.cs [hunk 2]
```diff
@@ -14,13 +14,27 @@ internal sealed class NegotiateSSPIContextProvider : SSPIContextProvider
...
-            SqlClientEventSource.Log.TryTraceEvent("TdsParserStateObjectManaged.GenerateSspiClientContext | Info | Session Id {0}, StatusCode={1}", _physicalStateObj.SessionId, statusCode);
+            NegotiateAuthenticationStatusCode statusCode = NegotiateAuthenticationStatusCode.UnknownCredentials;
+
+            for (int i = 0; i < _sniSpnBuffer.Length; i++)
+            {
+                string spnName = Encoding.Unicode.GetString(_sniSpnBuffer[i]);
+                _negotiateAuth ??= new(new NegotiateAuthenticationClientOptions { Package = "Negotiate", TargetName = spnName });
+                sendBuff = _negotiateAuth.GetOutgoingBlob(received.Span, out statusCode)!;
+                // Log session id, status code and the actual SPN used in the negotiation
+                SqlClientEventSource.Log.TryTraceEvent($"TdsParserStateObjectManaged.GenerateSspiClientContext | Info | Session Id {_physicalStateObj.SessionId}, StatusCode={statusCode}, SPN={_negotiateAuth.TargetName}");
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SSPI/NegotiateSSPIContextProvider.cs:25]: The previous style of `TryTraceEvent` has internal null checking and it's safer than this style in addition to the higher perf when trace is not enabled.
```suggestion
                SqlClientEventSource.Log.TryTraceEvent("TdsParserStateObjectManaged.GenerateSspiClientContext | Info | Session Id {0}, StatusCode={1}, SPN={2}", _physicalStateObj.SessionId, statusCode, spnName);
```

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SSPI/NegotiateSSPIContextProvider.cs [hunk 3]
```diff
@@ -14,13 +14,27 @@ internal sealed class NegotiateSSPIContextProvider : SSPIContextProvider
...
         {
-            _negotiateAuth ??= new(new NegotiateAuthenticationClientOptions { Package = "Negotiate", TargetName = Encoding.Unicode.GetString(_sniSpnBuffer[0]) });
-            sendBuff = _negotiateAuth.GetOutgoingBlob(received.Span, out NegotiateAuthenticationStatusCode statusCode)!;
-            SqlClientEventSource.Log.TryTraceEvent("TdsParserStateObjectManaged.GenerateSspiClientContext | Info | Session Id {0}, StatusCode={1}", _physicalStateObj.SessionId, statusCode);
+            NegotiateAuthenticationStatusCode statusCode = NegotiateAuthenticationStatusCode.UnknownCredentials;
+
+            for (int i = 0; i < _sniSpnBuffer.Length; i++)
+            {
+                string spnName = Encoding.Unicode.GetString(_sniSpnBuffer[i]);
+                _negotiateAuth ??= new(new NegotiateAuthenticationClientOptions { Package = "Negotiate", TargetName = spnName });
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SSPI/NegotiateSSPIContextProvider.cs:22]: If you don't tend to use this variable, then it can be removed.
```suggestion
                _negotiateAuth ??= new(new NegotiateAuthenticationClientOptions { Package = "Negotiate", TargetName = Encoding.Unicode.GetString(_sniSpnBuffer[i]) });
```

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SSPI/NegotiateSSPIContextProvider.cs (hunk 2 above)
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SSPI/NegotiateSSPIContextProvider.cs:24]: It seems you accidentally reverted this change!

**Review by DavoudEshtehari — APPROVED** [review]
